ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -485,6 +489,118 @@ object KeyValueCommitting {
 
   // ------------------------------------------------------
 
+  // Helper to lookup contract instances. We verify the activeness of
+  // contract instances here. Since we look up every contract that was
+  // an input to a transaction, we do not need to verify the inputs separately.
+  private def lookupContract(
```

Comment:
Do you think the following refactoring makes sense? Running the tests found no issue.

```scala
  private def lookupContract(
      inputState: Map[DamlStateKey, Option[DamlStateValue]],
      inputLogEntries: Map[DamlLogEntryId, DamlLogEntry],
      submitter: String,
      txLet: Timestamp,
      coid: AbsoluteContractId) = {
    def isVisibleToSubmitter(cs: DamlContractState): Boolean =
      cs.getLocallyDisclosedToList.asScala.contains(submitter) || cs.getDivulgedToList.asScala
        .contains(submitter) || {
        logger.trace(s"lookupContract($coid): Contract state not found!")
        false
      }
    def isActive(cs: DamlContractState): Boolean = {
      val activeAt = Option(cs.getActiveAt).map(parseTimestamp)
      activeAt.exists(txLet >= _) || {
        val activeAtStr = activeAt.fold("<activeAt missing>")(_.toString)
        logger.trace(s"lookupContract($coid): Contract not active (let=$txLet, activeAt=$activeAtStr).")
        false        
      }
    }      
    val (eid, nid) = absoluteContractIdToLogEntryId(coid)
    val stateKey = absoluteContractIdToStateKey(coid)
    for {
      // Fetch the state of the contract so that activeness and visibility can be checked.
      contractState <- inputState.get(stateKey).flatMap(_.map(_.getContractState)).orElse {
        logger.trace(s"lookupContract($coid): Contract state not found!")
        throw Err.MissingInputState(stateKey)
      }
      if isVisibleToSubmitter(contractState) && isActive(contractState)
      // Finally lookup the log entry containing the create node and the contract instance.
      entry = inputLogEntries.getOrElse(eid, throw Err.MissingInputLogEntry(eid))
      contract <- lookupContractInstanceFromLogEntry(eid, entry, nid)
    } yield contract
  }
```

Reply:
thanks, looks better.